src/NuGet.Core/NuGet.Common/ActivityCorrelationContext.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+#if DNXCORE50
+using System.Threading;
+#else
+using System.Runtime.Remoting.Messaging;
+#endif
+
+namespace NuGet.Common
+{
+    /// <summary>
+    /// Ambient context contains properties pertaining to a current activity.
+    /// Single activity engages multiple method calls at different layers. 
+    /// Sometimes it's necessary to identify separate calls belonging to the same activity if shared state is needed.
+    /// </summary>
+    public class ActivityCorrelationContext : IDisposable
+    {
+        private static readonly string DataSlotId = "NuGet.ActivityCorrelationContext";
+
+#if DNXCORE50
+        private static readonly AsyncLocal<ActivityCorrelationContext> _instance = new AsyncLocal<ActivityCorrelationContext>();
+#endif
+
+        private static readonly ActivityCorrelationContext _globalInstance = new ActivityCorrelationContext
+        {
+            CorrelationId = default(Guid).ToString()
+        };
+
+        /// <summary>
+        /// Represents activity unique ID.
+        /// </summary>
+        public string CorrelationId { get; private set; }
+
+        /// <summary>
+        /// Returns ambient context value or global instance if not set earlier.
+        /// </summary>
+        public static ActivityCorrelationContext Current
+        {
+            get
+            {
+#if DNXCORE50
+                return _instance.Value ?? _globalInstance;
+#else
+                var context = CallContext.LogicalGetData(DataSlotId);
```

Comment:
Please keep the code depending on 4.5 and not 4.6

From: Alex Panov [mailto:notifications@github.com]
Sent: Wednesday, April 13, 2016 12:58 PM
To: NuGet/NuGet.Client <NuGet.Client@noreply.github.com>
Cc: Yishai Galatzer <yigalatz@microsoft.com>
Subject: Re: [NuGet/NuGet.Client] Properly handle cancelled-by-user response in HttpSource (#388)


In src/NuGet.Core/NuGet.Common/ActivityCorrelationContext.cs<https://github.com/NuGet/NuGet.Client/pull/388#discussion_r59614494>:

> +        /// <summary>

> +        /// Represents activity unique ID.

> +        /// </summary>

> +        public string CorrelationId { get; private set; }

> +

> +        /// <summary>

> +        /// Returns ambient context value or global instance if not set earlier.

> +        /// </summary>

> +        public static ActivityCorrelationContext Current

> +        {

> +            get

> +            {

> +#if DNXCORE50

> +                return _instance.Value ?? _globalInstance;

> +#else

> +                var context = CallContext.LogicalGetData(DataSlotId);

AsyncLocal<T> is a new class available since .NET 4.6. As the fix initially targeted 3.4.x release I had to provide alternative solution using good old CallContext.

I'll go ahead and re-write it to use AsyncLocal everywhere for dev.

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub<https://github.com/NuGet/NuGet.Client/pull/388/files/796f4a979ba04ea6ca36364c77041996ff19e067#r59614494>